src/Conversion/ONNXToKrnl/Tensor/OneHot.cpp

```diff
@@ -0,0 +1,189 @@
+//===---------------- OneHot.cpp - Lowering OneHot Op -------------------===//
+//
+// This file lowers the ONNX OneHot Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+struct ONNXOneHotOpLowering : public ConversionPattern {
+  ONNXOneHotOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXOneHotOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXOneHotOpAdaptor operandAdaptor(operands);
+    ONNXOneHotOp oneHotOp = llvm::cast<ONNXOneHotOp>(op);
+    auto loc = op->getLoc();
+
+    Value indices = operandAdaptor.indices();
+    Value depth = operandAdaptor.depth();
+    Value values = operandAdaptor.values();
+
+    auto depthShape = depth.getType().cast<MemRefType>().getShape();
+    auto indicesShape = indices.getType().cast<MemRefType>().getShape();
+    int64_t indicesRank = indicesShape.size();
+
+    // Axis is a required attribute and should have default value of -1.
+    int64_t axisValue = oneHotOp.axis();
+    assert(axisValue >= 0 && axisValue <= indicesRank);
+
+    // Insert an allocation and deallocation for the result of this operation.
+    auto memRefOutType = convertToMemRefType(*op->result_type_begin());
+    auto memRefOutShape = memRefOutType.getShape();
+    int64_t outRank = memRefOutType.getRank();
+
+    IndexExprScope ieScope(rewriter, loc);
+    DimsExpr outputDims(outRank);
+    MemRefBoundsIndexCapture indicesBounds(indices);
+
+    // Allocate result.
+    Value alloc;
+    Value zero = emitConstantOp(rewriter, loc, rewriter.getIndexType(), 0);
+    Value one = emitConstantOp(rewriter, loc, rewriter.getIndexType(), 1);
+
+    // Load values depending on shape.
+    Value loadedDepth;
+    if (depthShape.size() == 0)
+      loadedDepth = rewriter.create<KrnlLoadOp>(loc, depth, ArrayRef<Value>{});
+    else if (depthShape.size() == 1 && depthShape[0] == 1)
+      loadedDepth = rewriter.create<KrnlLoadOp>(loc, depth, zero);
+    else
+      llvm_unreachable("depth shape must be 0 or if 1, size must be 1");
+
+    // convert float type to int64 for depth value
+    Type depthElmentType =
+        depth.getType().dyn_cast<MemRefType>().getElementType();
+
+    TypeSwitch<Type>(depthElmentType)
```

Comment:
The code for depth is correct but could be improved. The IE is supposed to be able to optimize the case when depth is a constant. Since there is a type cast here, I am not sure. Best we can get precise shape when depth is constant. In ONNXShapeHelper:1510 line, aixDim is given a ? without any attempt to optimize. Is that right?
We do not have to optimize it if it is too difficult.

Reply:
In ONNXShapeHelper, aixDim is given a `?` first and then assign depth value if it is constant. So we can infer shape when depth is constant.
```
%depth = "onnx.Constant"() {value = dense<10> : tensor<i64>} : () -> tensor<i64>
%0 = "onnx.OneHot"(%arg0, %depth, %arg1) : (tensor<2x2xi64>, tensor<i64>, tensor<2xf32>) -> tensor<*xf32>
  return %0 : tensor<*xf32>
.....
// CHECK: {{.*}} = "onnx.OneHot"(%arg0, [[R0]], %arg1) {axis = 2 : si64} : (tensor<2x2xi64>, tensor<i64>, tensor<2xf32>) -> tensor<2x2x10xf32>
```

And we get an unknown dimension when depth needs a input
```
%0 = "onnx.OneHot"(%arg0, %arg1, %arg2) : (tensor<2x2xi64>, tensor<i64>, tensor<2xf32>) -> tensor<*xf32>
  return %0 : tensor<*xf32>
......
// CHECK: {{.*}} = "onnx.OneHot"(%arg0, %arg1, %arg2)  {axis = 2 : si64} : (tensor<2x2xi64>, tensor<i64>, tensor<2xf32>) -> tensor<2x2x?xf32>
```

Do you mean that during ONNXOneHotOp lowering, we do not need type cast here if depth is already a constant?

Reply:
I see your following comment...